From f3417eb98cff4bad0dbfbb7b5d3a0d69f8a10106 Mon Sep 17 00:00:00 2001
From: Allison Vacanti <allison.vacanti@kitware.com>
Date: Tue, 12 Sep 2017 09:35:38 -0400
Subject: Stabilize depth sort to ensure consistent results.

Differences in unstable sorting algorithms led to inconsistent results
across platforms for the VTK project. This patch ensures that primitives
will be sorted the same everywhere.
---
 gl2ps.c | 15 +++++++++++++--
 1 file changed, 13 insertions(+), 2 deletions(-)

diff --git a/gl2ps.c b/gl2ps.c
index 6a957e4..1dcb617 100644
--- a/gl2ps.c
+++ b/gl2ps.c
@@ -164,7 +164,7 @@ typedef struct {
   GLshort type, numverts;
   GLushort pattern;
   char boundary, offset, culled;
-  GLint factor, linecap, linejoin;
+  GLint factor, linecap, linejoin, sortid;
   GLfloat width, ofactor, ounits;
   GL2PSvertex *verts;
   union {
@@ -600,6 +600,15 @@ static void gl2psListSort(GL2PSlist *list,
   qsort(list->array, list->n, list->size, fcmp);
 }
 
+/* Must be a list of GL2PSprimitives. */
+static void gl2psListAssignSortIds(GL2PSlist *list)
+{
+  GLint i;
+  for(i = 0; i < gl2psListNbr(list); i++){
+    (*(GL2PSprimitive**)gl2psListPointer(list, i))->sortid = i;
+  }
+}
+
 static void gl2psListAction(GL2PSlist *list, void (*action)(void *data))
 {
   GLint i;
@@ -1458,7 +1467,8 @@ static int gl2psCompareDepth(const void *a, const void *b)
     return 1;
   }
   else{
-    return 0;
+    /* Ensure that initial ordering is preserved when depths match. */
+    return q->sortid < w->sortid ? -1 : 1;
   }
 }
 
@@ -5885,6 +5895,7 @@ static GLint gl2psPrintPrimitives(void)
     gl2psListReset(gl2ps->primitives);
     break;
   case GL2PS_SIMPLE_SORT :
+    gl2psListAssignSortIds(gl2ps->primitives);
     gl2psListSort(gl2ps->primitives, gl2psCompareDepth);
     if(gl2ps->options & GL2PS_OCCLUSION_CULL){
       gl2psListActionInverse(gl2ps->primitives, gl2psAddInImageTree);
-- 
2.23.0

